feat: HUMSIG-37: ML backend experience updates

niklub opened the pull request:
- Support for HTTP Basic Authentication in ML Backend and Label Studio
- Capability to transfer additional JSON params from Label Studio to the ML Backend
- Enhanced configuration method for selecting predictions or ML Models in Label Studio for prelabelling
- Project settings featuring a list of predictions with options to remove them
- Standard screens for scenarios without connected models or uploaded predictions

## Old interface
![image](https://github.com/HumanSignal/label-studio/assets/6087484/bd27ad0a-d281-4fdc-bb4b-9d44baaf7423)
![image](https://github.com/HumanSignal/label-studio/assets/6087484/57d86bb3-24e1-406f-aa2e-2631519d2459)


## New interface
![image](https://github.com/HumanSignal/label-studio/assets/6087484/526b86a1-bbcb-41ff-982c-7ca99f595c37)
![image](https://github.com/HumanSignal/label-studio/assets/6087484/e18fab2b-2b3c-4dbe-954e-93495450b107)

Commit:
Updated implementation of the ML Backend experience. Docs & Test to follow

Commit:
Updating the experience a bit more based on the feedback. Plus
updating tests

Commit:
minor updates on the experience

Commit:
removing debug info, uncommenting bits

deppp committed:
Merge branch 'develop' into dev/ml-backend-exp

Commit:
fixing small issue to make it backward compatible with the previous project setting

Commit:
Merge branch 'dev/ml-backend-exp' of https://github.com/heartexlabs/label-studio into dev/ml-backend-exp

Commit:
Fix errors, code cleanup, fix ruff

Commit:
Downgrade testing-library, reformat with linters

Commit:
Remove excessive calls in task api

Commit:
Fix frontend script

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8192397383

jombooth committed:
fmt

Commit:
fix sdk version & change api /predict/test

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8201028879

Commit:
try update lock

bmartel committed:
Merge branch 'develop' into 'fb-ml-backend-exp'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8207748036

bmartel committed:
[submodules] Copy src HumanSignal/dm2

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8207748036

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8207796735

Commit:
Address review comments with stylistic changes, remove unusable code, change some writings

Commit:
Additional changes

Commit:
Add svg icon file

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8217959028

Commit:
Fix ml/predict/test api

bmartel committed:
Running a formatter/lint on the code

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8237179655

bmartel committed:
Running a formatter/lint on the code

bmartel committed:
Merge remote-tracking branch 'origin/fb-ml-backend-exp' into fb-ml-backend-exp

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8237277894

bmartel committed:
Running a formatter/lint on the code

bmartel committed:
Merge remote-tracking branch 'origin/fb-ml-backend-exp' into fb-ml-backend-exp

bmartel committed:
Running a formatter/lint on the code

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8237418085

Commit:
Handle security on password input

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8238405264

Commit:
Display default password when not specified

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8240219216

Commit:
Reduce number of calls due to project id cache

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8256398303

bmartel commented on label_studio/data_manager/actions/basic.py:
```diff
@@ -138,6 +138,7 @@ def async_project_summary_recalculation(tasks_ids_list, project_id):
         'title': 'Retrieve Predictions',
         'order': 90,
         'dialog': {
+            'modal_title': 'Retrieve Predictions',
```
```suggestion
            'title': 'Retrieve Predictions',
```

bmartel commented on label_studio/data_manager/actions/predictions_to_annotations.py:
```diff
@@ -95,8 +95,10 @@ def predictions_to_annotations_form(user, project):
         'dialog': {
-            'text': 'This action will create new annotations from predictions with the selected model version '
-            'for each selected task.',
+            'modal_title': 'Create Annotations From Predictions',
```
```suggestion
            'title': 'Create Annotations From Predictions',
```

Commit:
Revert pyproject.toml

Commit:
Run poetry lock --no-update

niklub committed:
Update label_studio/data_manager/actions/predictions_to_annotations.py

Co-authored-by: bmartel <brandonmartel@gmail.com>

niklub committed:
Update label_studio/data_manager/actions/basic.py

Co-authored-by: bmartel <brandonmartel@gmail.com>

Commit:
Run poetry lock --no-update after revert to develop

jombooth commented on label_studio/tasks/api.py:
```diff
@@ -223,20 +225,21 @@ def get_serializer_class(self):
-        # call machine learning api and format response
-        if project.evaluate_predictions_automatically:
-            for ml_backend in task.project.ml_backends.all():
-                ml_backend.predict_tasks([task])
+    #     # call machine learning api and format response
+    #     if project.retrieve_predictions_automatically:
+    #         for ml_backend in task.project.ml_backends.all():
+    #             ml_backend.predict_tasks([task])
 
-        result = self.get_serializer(task).data
+    #     result = self.get_serializer(task).data
 
-        # use proxy inlining to task data (for credential access)
-        result['data'] = task.resolve_uri(result['data'], project)
-        return Response(result)
+    #     # use proxy inlining to task data (for credential access)
+    #     result['data'] = task.resolve_uri(result['data'], project)
+    #     return Response(result)
```
why is this commented out? If it's no longer needed, let's delete it. But we should understand what it was doing before

jombooth commented on label_studio/tasks/api.py:
```diff
@@ -181,18 +180,21 @@ def get_retrieve_serializer_context(self, request):
+        context['project'] = self.task.project
 
         # get prediction
-        if (
-            project.evaluate_predictions_automatically or project.show_collab_predictions
-        ) and not self.task.predictions.exists():
```
Why do we no longer check for `project.evaluate_predictions_automatically`?

jombooth replied:
I have since learned that it was deprecated as part of this work - we should add this fact to the PR description

jombooth commented on label_studio/tasks/api.py:
```diff
@@ -181,18 +180,21 @@ def get_retrieve_serializer_context(self, request):
+            self.task.refresh_predictions()
             self.task.refresh_from_db()
 
+        # predictions = retrieve_predictions([task])
+        # if predictions:
+        #     self.task.refresh_from_db()
```
this looks like it can be deleted

jombooth commented on .pre-commit-dev.yaml:
```diff
@@ -9,3 +9,4 @@ repos:
     rev: v0.9.1
     hooks:
       - id: blue
+        args: [ --verbose ]
```
please revert this, it spams my terminal when I run `make fmt-all`

jombooth commented on label_studio/data_manager/api.py:
```diff
@@ -255,11 +255,15 @@ def get(self, request):
+            # people to retrieve manually instead on DM load, plus it
+            # will slow down initial DM load
+
             # retrieve ML predictions if tasks don't have them
-            if not review and project.evaluate_predictions_automatically:
-                tasks_for_predictions = Task.objects.filter(id__in=ids, predictions__isnull=True)
-                evaluate_predictions(tasks_for_predictions)
-                [tasks_by_ids[_id].refresh_from_db() for _id in ids]
+            # if not review and project.retrieve_predictions_automatically:
+            #     tasks_for_predictions = Task.objects.filter(id__in=ids, predictions__isnull=True)
+            #     retrieve_predictions(tasks_for_predictions)
+            #     [tasks_by_ids[_id].refresh_from_db() for _id in ids]
```
What's the outcome of the discussion here? looks like another place where we used to consider `project.retrieve_predictions_automatically` but don't anymore. In the final version, we shouldn't have commented out code here.

Commit:
Remove unused comments

jombooth commented on label_studio/data_manager/api.py:
```diff
@@ -271,13 +275,18 @@ def get(self, request):
-        if project.evaluate_predictions_automatically:
-            evaluate_predictions(queryset.filter(predictions__isnull=True))
+        # if project.retrieve_predictions_automatically:
+        #     retrieve_predictions(queryset.filter(predictions__isnull=True))
```
Similarly, we shouldn't merge commented out code here. What's the intent of this part of the change?

jombooth commented on label_studio/ml/api.py:
```diff
@@ -227,29 +281,64 @@ def post(self, request, *args, **kwargs):
     ),
 )
 class MLBackendInteractiveAnnotating(APIView):
+    """ """
```
We can delete this, or better yet, add a docstring :)

jombooth replied:
(applies to all the `""" """` docstrings below also)

jombooth commented on label_studio/tests/predictions/test_predictions.py:
```diff
@@ -769,6 +769,8 @@ def test_interactive_annotating_failing(business_client, configured_project):
 
         result = r.json()
 
+    print(result)
+
```
please remove this print

jombooth commented on label_studio/data_manager/functions.py:
```diff
@@ -314,16 +314,19 @@ def get_prepared_queryset(request, project):
-    project = tasks[0].project
+    if not backend:
+        project = tasks[0].project
+        backend = project.ml_backends.first()
```
this is a small table so it likely doesn't matter, but it's generally better to use `fast_first` if you don't care about ordering

niklub replied:
since it is only one backend connected all the time, I wouldn't introduce this change rn to avoid big diff

jombooth commented on label_studio/data_manager/functions.py:
```diff
@@ -314,16 +314,19 @@ def get_prepared_queryset(request, project):
-    for ml_backend in project.ml_backends.all():
-        # tasks = tasks.filter(~Q(predictions__model_version=ml_backend.model_version))
-        ml_backend.predict_tasks(tasks)
+    # IMPORTANT change here, ml_backends.all => ml_backends.first
+    # we are using only one ML backend, not multiple
```
I think this belongs in the docstring. What I'd recommend:

```
"""
Call the given ML backend to retrieve predictions with the task queryset as an input.
If backend is not specified, we'll assume the tasks' project only has one associated
ML backend, and use that backend.
"""
```

jombooth commented on label_studio/ml/api.py:
```diff
@@ -202,6 +204,58 @@ def post(self, request, *args, **kwargs):
+            kwargs = ml_backend._predict(task)
+            return Response(**kwargs)
+
+        # TODO this needs to be implemented and needs to have a specific task param
```
Is this still a TODO? I can't really tell what's meant by the "this"

jombooth commented on label_studio/tasks/models.py:
```diff
@@ -178,6 +179,20 @@ class Meta:
+    @classmethod
+    def get_random(cls, project):
+        """Get random task from a project, this should not be used lightly as its expensive method to run"""
+        if project is None:
+            raise Exception('Project id is required')
```
The method shouldn't raise an exception for this. `project` should be a required argument, and there shouldn't be any nullchecking.

jombooth commented on label_studio/ml/models.py:
```diff
@@ -391,6 +529,16 @@ def _validate_ml_api_result(ml_api_result, tasks, curr_logger):
     return True
 
 
+@receiver(pre_delete, sender=MLBackend)
+def modify_project_model_version(sender, instance, **kwargs):
+    """ """
```
We generally discourage the use of Django signals because they aren't called under various conditions, eg, a queryset-level delete. why is this necessary? A docstring would really help.

jombooth commented on label_studio/ml/models.py:
```diff
@@ -391,6 +529,16 @@ def _validate_ml_api_result(ml_api_result, tasks, curr_logger):
+
+    if project.model_version == instance.title:
+        project.model_version = None
+        project.save()
```
please use update_fields to avoid the potential for clobbering other updates to `project`. like `project.save(update_fields=['model_version'])`

jombooth commented on label_studio/ml/models.py:
```diff
@@ -87,45 +123,110 @@ class MLBackend(models.Model):
+        if self.title != self.__original_title and p.model_version == self.__original_title:
+            with transaction.atomic():
+                p.model_version = self.title
+                p.save()
```
this should also use update_fields as recommended in https://github.com/HumanSignal/label-studio/pull/5570/files#r1522245983

jombooth commented on label_studio/ml/models.py:
```diff
@@ -87,45 +123,110 @@ class MLBackend(models.Model):
+        super(MLBackend, self).__init__(*args, **kwargs)
+        self.__original_title = self.title
+
+    def save(self, *args, **kwargs):
+        """
+        Overrides the save() method to update the associated project's model_version field.
+
+        If the title of the model instance is changed and the model_version
+        of the related project is currently the same as the original title,
+        the project's model_version is updated to the new title.
+        """
```
It's really better to avoid custom `save()` methods if at all possible. Is there a different way we could do this? For example, could we just make the project's model_version optional and defer to the title of the model instance if it is unset?

jombooth commented on label_studio/ml/models.py:
```diff
@@ -87,45 +123,110 @@ class MLBackend(models.Model):
+        )
+
     def update_state(self):
+        """ """
```
nit: please add a docstring or remove this

jombooth commented on label_studio/ml/models.py:
```diff
@@ -30,6 +30,11 @@ class MLBackendState(models.TextChoices):
 
 
+class MLBackendAuth(models.TextChoices):
+    NONE = 'NA', _('None')
+    BASIC_AUTH = 'BA', _('Basic Auth')
```
We save almost nothing by using two-letter codes in this charfield, why not make the CharField len 255 and make the enum values the same as the enum name? This means nobody ever has to figure out that "NA" is the same as NONE by reading backend code, for example.

jombooth commented on label_studio/ml/models.py:
```diff
@@ -60,20 +65,51 @@ class MLBackend(models.Model):
+    basic_auth_user = models.TextField(
+        _('basic auth user'),
+        blank=True,
+        null=True,
+        default='',
```
On recent versions of postgres, adding a column with a simple default is an efficient operation. I would have preferred specifying either `null=True` or `default=''` but not both; however I recognize the ship may have already sailed on this since it was deployed on a production database.

jombooth commented on label_studio/ml/serializers.py:
```diff
@@ -2,50 +2,119 @@
+            'error_message',
+            'title',
+            'auth_method',
+            'basic_auth_user',
+            'basic_auth_pass',
```
We need to make sure that these fields are writable but not readable in the API.

jombooth replied:
ok, I do see this is write_only: https://github.com/HumanSignal/label-studio/pull/5570/files#diff-a67b82b0de34ce236f0f9a212e530d2a865c73f603e0e5e6ca485d5c0ea45d10R13

would be great to capture this in a test though.

jombooth commented on label_studio/projects/models.py:
```diff
@@ -259,12 +271,17 @@ class SkipQueue(models.TextChoices):
         self.__overlap_cohort_percentage = self.overlap_cohort_percentage
         self.__skip_queue = self.skip_queue
 
+        # if rpa is not None:
+        #     self.evaluate_predictions_automatically = rpa
```
need to remove this too

jombooth commented on label_studio/projects/models.py:
```diff
@@ -276,18 +293,19 @@ def __init__(self, *args, **kwargs):
-            return Prediction.objects.filter(Q(project=self.id) & Q(model_version=self.model_version))
-        else:
-            return Prediction.objects.filter(Q(task__project=self.id) & Q(model_version=self.model_version))
+    @property
+    def ml_backend(self):
+        return self.ml_backends.first()
```
please use `fast_first` instead of `first()`

jombooth commented on label_studio/projects/models.py:
```diff
@@ -644,16 +658,41 @@ def display_count(count: int, type: str) -> Optional[str]:
+
+            predictions.delete()
+
+        ## TODO wtf is that returning
```
We should remove this comment

jombooth commented on label_studio/projects/models.py:
```diff
@@ -644,16 +658,41 @@ def display_count(count: int, type: str) -> Optional[str]:
+            # If we are deleting specific model_version then we need
+            # to remove that from the project
+            if self.should_none_model_version(model_version):
+                self.model_version = None
+                self.save()
```
we should be using update_fields here too

jombooth commented on label_studio/projects/models.py:
```diff
@@ -644,16 +658,41 @@ def display_count(count: int, type: str) -> Optional[str]:
+                self.model_version = None
+                self.save()
+
+            predictions.delete()
```
optional: I believe the delete() method returns the number of rows that it deleted, so we can ditch the count() query

jombooth commented on label_studio/projects/serializers.py:
```diff
@@ -166,6 +167,36 @@ def validate_label_config(self, value):
+
+    def update(self, instance, validated_data):
+        """ """
+        # assume there is a flag, 'is_name_updated', in model to track if a name has been updated
```
This comment seems to be outdated, or is `'show_collab_predictions'` wrong below?

jombooth commented on label_studio/tests/ml/test_models.py:
```diff
@@ -0,0 +1 @@
+
```
please remove this and other empty files from the diff

jombooth commented on label_studio/projects/api.py:
```diff
@@ -640,5 +642,34 @@ class ProjectModelVersions(generics.RetrieveAPIView):
     queryset = Project.objects.all()
 
     def get(self, request, *args, **kwargs):
+        # TODO make sure "extended" is the right word and is
+        # consistent with other APIs we've got
```
I think this should be addressed before merging

niklub replied:
I think we don't have consistent naming convention regarding request args

jombooth commented on label_studio/ml/api_connector.py:
```diff
@@ -31,18 +31,37 @@
         self._headers = headers or {}
+        self._auth_method = auth_method
+
+        if auth_method == 'BA':
```
Here and everywhere in the PR: we shouldn't refer to strings from an enum explicitly. Instead, let's compare `auth_method` with `MLBackendAuth.BASIC_AUTH`

Ideally we would use an enum on the frontend too for the same reason.

jombooth commented on label_studio/projects/models.py:
```diff
@@ -644,16 +658,41 @@ def display_count(count: int, type: str) -> Optional[str]:
+        """
+        return self.model_version == model_version or self.ml_backend_in_model_version
+
+    def delete_predictions(self, model_version=None, *args, **kwargs):
```
I don't see `*args` or `**kwargs` being used anywhere

niklub commented on label_studio/ml/models.py:
```diff
@@ -136,13 +237,12 @@ def update_state(self):
             else:
                 self.state = MLBackendState.CONNECTED
-                model_version = setup_response.response.get('model_version')
+                model_version = setup_response.response.get('model_version', 'undefined')
```
let's fix this

niklub commented on label_studio/ml/models.py:
```diff
@@ -156,52 +256,142 @@ def train(self):
+            },
+        }
+
+    def predict_and_save(self, tasks=None):
```
maybe rollack it again

niklub commented on label_studio/ml/serializers.py:
```diff
@@ -2,50 +2,119 @@
+    """ """
+
+    readable_state = serializers.SerializerMethodField()
+    basic_auth_pass = serializers.CharField(write_only=True, required=False, allow_null=True, allow_blank=True)
```
adding test to check GET payload

Commit:
Code revert cleanup, revert tavern tests

Commit:
Remove tavern tests

Commit:
REplace strings in MLBackendAuth

jombooth commented on label_studio/projects/serializers.py:
```diff
@@ -166,6 +166,34 @@ def validate_label_config(self, value):
+            # or match version in predictions
+
+            if (
+                p.ml_backends.filter(title=value).exists()
+                or p.predictions.filter(project=p, model_version=value).exists()
```
I think you could do (pseudocode)

```
MlBackend.objects.filter(project=p, title=value).union(Prediction.objects.filter(project=p, model_version=value)).exists()
```

this would be one query instead of four, assuming p.ml_backends and p.predictions aren't prefetched.

Commit:
Update migrations

jombooth commented on label_studio/tests/next_task.tavern.yml:
```diff
@@ -116,31 +97,6 @@ stages:
     json:
       data:
         text: Test example phrase
-      predictions:
```
what's behind the changes to this test?

Commit:
Revert back predictions tavern

jombooth committed:
redo migrations without two-letter abbrevs

jombooth committed:
fmt

jombooth committed:
fix the frontend to match

jombooth committed:
remove new empty files and empty docstrings

jombooth committed:
use fast_first, DRY

jombooth committed:
revert test refactorings

jombooth committed:
fix bug and remove unnecessary query

jombooth committed:
more test unrefactoring

jombooth committed:
merge develop

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8287275155

jombooth committed:
fix some test issues

jombooth commented on label_studio/ml/api.py:
```diff
@@ -227,29 +283,65 @@ def post(self, request, *args, **kwargs):
+        if ml_api_result.is_error:
+            message = f'Prediction not created for project {self}: {ml_api_result.error_message}'
+            return self._error_response(message)
+
+        if not isinstance(ml_api_result.response, dict) or 'results' not in ml_api_result.response:
+            message = f'Incorrect response from ML service it must be a dict and contain "results" key: {ml_api_result.response}'
+            return self._error_response(message)
+
+        ml_results = self._get_ml_results(ml_api_result)
 
-        result = ml_backend.interactive_annotating(task, context, user=request.user)
+        if not ml_results:
+            message = f'ML backend has to return a list with at least 1 annotation but it returned: {type(ml_results)}'
+            return self._error_response(message, logger.warning)
 
-        return Response(
-            result,
-            status=status.HTTP_200_OK,
-        )
+        return Response({'data': ml_results}, status=status.HTTP_200_OK)
```
There are three test failures coming from this code, because `ml_api_result` at the top is a dictionary, not an `MLApiResult`, and therefore doesn't have any members like `is_error`. The tests that need to be fixed are:

```
tests/test_predictions.py::test_interactive_annotating
tests/test_predictions.py::test_interactive_annotating_failing
tests/test_predictions.py::test_interactive_annotating_with_drafts
```

Commit:
Fix interactive api

jombooth commented on label_studio/ml/models.py:
```diff
@@ -96,36 +131,76 @@ class MLBackend(models.Model):
+        super(MLBackend, self).__init__(*args, **kwargs)
+        self.__original_title = self.title
+
+    def save(self, *args, **kwargs):
```
I think to be comfortable merging this, we really need a testcase that captures the behavior of this function (ideally something relatively high-level demonstrating the sequence of realistic API calls that causes project.model_version to be renamed, not just a unit test)

jombooth commented on label_studio/ml/models.py:
```diff
@@ -391,6 +498,15 @@ def _validate_ml_api_result(ml_api_result, tasks, curr_logger):
+
+    if project.model_version == instance.title:
+        project.model_version = None
+        project.save(update_fields=['model_version'])
```
as with MLBackend#save above, we really need a testcase that, using realistic API calls, captures the impact that this signal handler has.

jombooth commented:
Requesting screenshots of all new in-app functionality in the description, two testcases for model_version update behavior on save and delete, and the GET api test for basic_auth_pass mentioned earlier in the review - with these changes, I'm prepared to approve for the backend

jombooth committed:
ensure accurate count of deleted predictions

Commit:
Add tests

jombooth commented on label_studio/projects/serializers.py:
```diff
@@ -177,8 +177,9 @@ def validate_model_version(self, value):
             if (
-                p.ml_backends.filter(title=value).exists()
-                or p.predictions.filter(project=p, model_version=value).exists()
+                # p.ml_backends.filter(title=value).exists()
+                # or p.predictions.filter(project=p, model_version=value).exists()
```
please remove commented out code

jombooth commented:
Thanks for adding old vs new screenshots, super helpful :+1:

Commit:
Add missing files, remove commented code

Commit:
Add resolve hostname mock in tests

Commit:
Mock hostname resolution fix

Commit:
Reformat

niklub committed:
Merge branch 'develop' into fb-HUMSIG-37

Commit:
Merge migrations

jombooth commented on label_studio/ml/api_connector.py:
```diff
@@ -31,18 +31,36 @@
         self._headers = headers or {}
+        self._auth_method = auth_method
+
+        # TODO add kwarg check in here
```
What does this mean? Please clarify.

jombooth reviewed (approved):
Approving for backend changes, please address remaining comments about TODOs etc

jombooth committed:
fix ruff linter issues

jombooth committed:
revert a change that isnt needed

bmartel reviewed (approved):
Approving for frontend. Thanks @niklub @jombooth 🎉

Commit:
Add todo comment

Commit:
Merge branch 'develop' into fb-HUMSIG-37

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/8348402932

Commit:
Add ml backend predict test api test to increase coverage

Commit:
Add more tests to increase coverage

jombooth committed:
fix ruff

jombooth reviewed (approved):
LGTM, also fixed ruff
